### ISSUE TITLE
Tilemap terrain not drawn when set via Task.Run 

### ISSUE THREAD

**Issue by andypandy47** [godotengine/godot#75317]
### Godot version

v4.0 stable mono

### System information

Windows 10, Vulkan

### Issue description

I've been playing around with noise to create a large procedural world where I'm generating and clearing chunks of tiles to help performance. I've found that there are sizeable framerate drops when setting and clearing terrain using set_cells_terrain_connect. These drops can be easily avoided by calling the function from a different thread in C# using Task.Run. However, occasionally the terrain will not generate properly. This seems to only be happening when called from within Task.Run.

In the screenshots below you can see where Godot has attempted to generate terrain and has been successful to an extent (see the collision shapes for the correct tiles) but patches of the tile sprites are seemingly not drawn.
<img width="593" alt="messed_up_terrain" src="https://user-images.githubusercontent.com/19389664/227715188-a44e3426-a10f-4f47-a598-3be61730f75e.png">
<img width="888" alt="messed_up_terrain_2" src="https://user-images.githubusercontent.com/19389664/227715191-1fa7b688-eab9-48e8-a90c-8a61da3ea6e8.png">

If I move the player far enough away to clear the tiles and then back again so that they are regenerated, the engine will generate the terrain fine. It seems to be very inconsistent about when it fails and there are no errors in the output.

### Steps to reproduce

Define a region within code then within a separate function, loop through the cells in the region adding each one to a new collection then use that collection as an arg to set the terrain using set_cells_terrain_connect. Repeatedly call this function as part of a Task.Run and eventually you will see incorrect terrain.

```
using Godot;
using System.Collections.Generic;
using System.Threading.Tasks;

public partial class GenerateTerrain : TileMap
{
	[Export]
	public Vector2I Region { get; set; }

	private FastNoiseLite Noise { get; set; } = new FastNoiseLite();

	private RandomNumberGenerator RandomNumberGenerator { get; set; } = new RandomNumberGenerator();

	// Called when the node enters the scene tree for the first time.
	public override void _Ready()
	{
		Task.Run(() => this.DrawTerrain(new Vector2I(50, 50)));
	}

	// Called every frame. 'delta' is the elapsed time since the previous frame.
	public override void _Process(double delta)
	{
		if (Godot.Input.IsMouseButtonPressed(MouseButton.Left))
		{
			this.Clear();
			var fromPos = this.LocalToMap(this.GetLocalMousePosition());
			Task.Run(() => this.DrawTerrain(fromPos));
		}
	}

	private Task DrawTerrain(Vector2I fromPos)
	{
		var startPoint = fromPos - this.Region / 2;
		var endPoint = fromPos + this.Region / 2;

		var groundTerrain = new List<Vector2I>();

		for(int x = startPoint.X; x < endPoint.X; x++)
		{
			for(int y = startPoint.Y; y < endPoint.Y; y++)
			{
				var cell = new Vector2I(x, y);

				groundTerrain.Add(cell);
			}
		}

		var groundTerrainCells = new Godot.Collections.Array<Vector2I>(groundTerrain);
		this.SetCellsTerrainConnect(0, groundTerrainCells, 0, 0, false);

		return Task.CompletedTask;
	}
}
```
I've attached a minimal reproduction project. Run the Test scene and click around the window. Each click should generate a new 50x50 square of terrain. Keep on clicking and eventually, you will see terrain being generated that is incorrect. See attached gif below
![dodgy_terrain_taskrun](https://user-images.githubusercontent.com/19389664/227716112-3e9c5e66-78da-4deb-8813-540f24be7302.gif)

Expected behaviour in the example project is to consistently generate a 50x50 square of terrain.

### Minimal reproduction project

[TerrainBug.zip](https://github.com/godotengine/godot/files/11068785/TerrainBug.zip)

**Comment by AThousandShips** [comment]
Or, wait, you are firing this several times at once, you're doing this each frame the mouse is held down, you need to do it just once when the mouse is pressed

**Comment by andypandy47** [comment]
Yeah good shout, I've changed the input now so it's `IsActionJustPressed` thats my bad, I'll update the reproduction project. The behaviour is still the same though. QueueRedraw does not seem to affect anything, unfortunately.  Awaiting the task and checking if it's still running will still produce the same effect.

```
public async override void _Process(double delta)
	{
		if (Godot.Input.IsActionJustPressed("Primary"))
		{
			if (this.DrawTerrainTask != null)
			{
				GD.Print("Task still running!");
				return;
			}

			this.Clear();
			var fromPos = this.LocalToMap(this.GetLocalMousePosition());
			this.DrawTerrainTask = Task.Run(() => this.DrawTerrain(fromPos));
			await this.DrawTerrainTask;
			this.DrawTerrainTask = null;
		}

		this.QueueRedraw();
	}
```

**Comment by AThousandShips** [comment]
Does it happen if you don't await?

**Comment by andypandy47** [comment]
Yeah still happens without awaiting

```
public override void _Process(double delta)
	{
		if (this.DrawTerrainTask != null && this.DrawTerrainTask.IsCompleted)
		{
			GD.Print("Task completed!");
			this.DrawTerrainTask = null;
		}

		if (Godot.Input.IsActionJustPressed("Primary"))
		{
			if (this.DrawTerrainTask != null)
			{
				GD.Print("Task still running!");
				return;
			}

			this.Clear();
			var fromPos = this.LocalToMap(this.GetLocalMousePosition());
			this.DrawTerrainTask = Task.Run(() => this.DrawTerrain(fromPos));
		}

		this.QueueRedraw();
	}
```
<img width="907" alt="image" src="https://user-images.githubusercontent.com/19389664/227717689-f6fdf0a4-faf8-4144-a55c-8d73a7675ae9.png">

**Comment by AThousandShips** [comment]
Then it returns to the original question, is it thread safe to update a tile map, doesn't seem like it, most of Godot isn't thread safe unless it's explicitly stated

**Comment by AThousandShips** [comment]
Setting a cell queues updating of the map, which since you're running on a thread can happen before you're done, and then won't fix it again as it's done, what you are doing is probably not thread safe

**Comment by AThousandShips** [comment]
What happens if you call `ForceUpdate()`

**Comment by andypandy47** [comment]
Ahhh I see, that was gonna be my next question lol, why is this particular thing not thread-safe? I'm not super knowledgeable in the world of multi-threaded programming but it seemed like giving it enough time to run the function between calls wouldn't cause any race conditions or anything. However, what you've said makes sense I think. I suppose I could have queued an update from one thread and then got overwritten by the main thread? something like that.

**Comment by andypandy47** [comment]
`ForceUpdate()` seems to work! Calling it within the task caused it to crash. Calling it from the process function at the point where it checks to see if the task is complete seems to make it work

```
public override void _Process(double delta)
	{
		if (this.DrawTerrainTask != null && this.DrawTerrainTask.IsCompleted)
		{
			GD.Print("Task completed!");
			this.DrawTerrainTask = null;
			this.ForceUpdate(0);
		}

		if (Godot.Input.IsActionJustPressed("Primary"))
		{
			if (this.DrawTerrainTask != null)
			{
				GD.Print("Task still running!");
				return;
			}

			this.Clear();
			var fromPos = this.LocalToMap(this.GetLocalMousePosition());
			this.DrawTerrainTask = Task.Run(() => this.DrawTerrain(fromPos));
		}
	}
```

**Comment by AThousandShips** [comment]
That forces it to recreate the textures and data, what probably happens is that it queues this update, and then marks that it's done, and won't do it again, because all the tiles changed between then are ignored

But simply most functionality isn't thread safe because it adds an overhead to do so, which makes it slower in all cases, I would suggest not using threading unless you can verify things are safe, and I'm not sure what you are doing is safe still, it could cause crashes in some cases or other weird bugs

**Comment by AThousandShips** [comment]
It *might* be safe to remove the map from the tree (unparent it) and do the update while not in the tree, because it won't do some of the updates then, but I'm not sure

**Comment by andypandy47** [comment]
Interesting. I'm going to try and implement it in my actual project and see what happens but if it keeps on causing issues will likely abandon the procedural generation idea or at least try to find a better method of chunking that won't cause such big frame drops. Thanks very much for your help and insight into Godot @AThousandShips

**Comment by AThousandShips** [comment]
No problem!

Consider tricks like separating the map into chunks which can be loaded before added to the tree, so you aren't affected by having to process it while not in the tree, that should be safe to do on a thread when I looked at the code (it won't do the update when outside the tree, and forces it to update when it enters, this should probably make things smoother as it won't wait for updates like that and won't refresh stuff until its ready)

**Comment by AThousandShips** [comment]
I think this can be closed unless there's some issue remaining, it seems to be related to thread unsafe issues

**Comment by Calinou** [comment]
> I think this can be closed unless there's some issue remaining, it seems to be related to thread unsafe issues

Is there something we could better document in the page you linked?

**Comment by AThousandShips** [comment]
Good question, I personally would say it makes it clear it's not thread safe, and suggests loading off the tree and then adding, but if someone else can chime in on that too, I'm relatively experienced on threading so might affect my perspective

**Comment by alberto-mco** [comment]
I reported a few month ago the same behavior in #71388, sometimes the tiles aren't painted. In my case, was using GDScript
